pkg/controller/common/helpers.go

```diff
@@ -125,6 +125,20 @@ func MergeMachineConfigs(configs []*mcfgv1.MachineConfig, osImageURL string) (*m
 		}
 	}
 
+	// For layering, we want to let the user override OSImageURL again
+	overriddenOSImageURL := ""
+	for _, cfg := range configs {
+		if cfg.Spec.OSImageURL != "" {
+			overriddenOSImageURL = cfg.Spec.OSImageURL
+		}
+	}
+	// Make sure it's obvious in the logs that it was overridden
+	if overriddenOSImageURL != "" && overriddenOSImageURL != osImageURL {
+
+		glog.Infof("OSImageURL has been overridden via machineconfig with %s (was: %s)", overriddenOSImageURL, osImageURL)
+		osImageURL = overriddenOSImageURL
+	}
+
```

Comment:
This is probably where I'd safety check/feature gate it -- just prevent the merge to keep the old behavior. 

If we aren't going to feature gate it, do we want a safety check in here that's:

1. check to make sure it's a "new format" image
2. only allow it to be overridden if we were able to check and it was a "new format" image?

So that no existing clusters would be impacted if they had, say, an old `MachineConfig` with an old image in it that previously would have been ignored?

Reply:
As we discussed, this does introduce the additional behaviour of having the controller pull the image as well (or at least a remote inspect), which isn't necessarily a bad thing, since if this errors, the daemon would likely have errored anyways, so we are surfacing this earlier in the upgrade pipeline.

To summarize our options then are:
1. add a featuregate to enable osImageURL overrides in 4.12, and then remove this in future versions once this is more mature (does featuregates prevent upgrades at all?)
2. add an additional image check in the controller with additional error
3. no additional safety at all. the daemon will error if you have an override image that isn't in the correct format, which would only trip up users in one scenario (the user had a osImageurl override from a previous version incorrectly, but since we didn't error, didn't even notice)

I think... I would lean towards 2 personally, but don't have a strong opinion. Essentially 2 and 3 are the "same" in terms of what would fail, but 2 surfaces it earlier, and 1 would just be "make sure you absolutely know what you are doing" which I would also be ok with so long as we have a plan out of it

Reply:
We talked on Monday and given:
-  the complexity of adding the safety 
-  some of the container/registry library constraints (potentially having to parse/manage `config.json` entries ourselves in the controller, or writing secrets out to disk) and
- the likelihood of someone actually having an old OSImageURL in their machineconfig

We opted not to put the safety in at this time, but we can later if we think we need to.